### PR TITLE
Addresses all CA1801 Code Analyzer issues

### PR DESCRIPTION
**Pull request by dominoFire**
## Bug

Fixes: https://github.com/NuGet/Client.Engineering/issues/640

Regression? Last working version: No

## Description

All CA1801 warnings are addressed using the following method:

```
For each Code Analyzer CA1801 warning, do the following:
1) if it is shipped Public API:
  1.a) Analyze if it's worth keeping it
  1.a) If yes, keep Code Analyzer warning, write justification and provide a xmldoc saying that parameter is not used.
2) otherwise:
  2.a) If it is a CancellationToken:
    2.a.i) either call ThrowIfCancellationRequested() or use as an argument inside the method body.
  2.b) otherwise:
    2.b.i) Verify no usage in commit history or multitargeting scenarios. If nothing is found, remove the parameter.
```

The following [spreadsheet](https://microsoft-my.sharepoint.com/:x:/p/feaguila/EQNs0t9FQohOoKtSvGlAZYoBD1-pw4FXAA2Dn7wLDgd2pw?e=AKjB8D) contains the actions taken for each warning listed in the issue.

## PR Checklist

- [x] PR has a meaningful title
- [x] PR has a linked issue.
- [x] Described changes

- **Tests**
  - [ ] Automated tests added
  - **OR**
  
  - [x] Test exception: Since there's no product change, no new tests are added. Let me know if I need to add test where cancellation tokens cases.
  - **OR**
  - [ ] N/A 

- **Documentation**
  
  - [x] Documentation PR or issue filled: https://github.com/NuGet/docs.microsoft.com-nuget/issues/2496
  - **OR**
  - [] N/A

### CODE REVIEW

File: src/NuGet.Clients/NuGet.CommandLine/Commands/ProjectFactory.cs [hunk 1]
```diff
@@ -64,6 +64,13 @@ public class ProjectFactory : MSBuildUser, IProjectFactory, CoreV2.NuGet.IProper
         [Import]
         public Configuration.IMachineWideSettings MachineWideSettings { get; set; }
 
+        /// <summary>
+        /// Creates a MSBuild project factory with data from NuGet <see cref="Commands.PackArgs"/> object
+        /// </summary>
+        /// <param name="packArgs">NuGet arguments to create a package</param>
+        /// <param name="path">Not used</param>
```
Comment (kartheekp-ms) [inline src/NuGet.Clients/NuGet.CommandLine/Commands/ProjectFactory.cs]: Given that this method is in `NuGet.CommandLine` project, I guess it should be okay to remove unused parameter.

File: src/NuGet.Clients/NuGet.CommandLine/GlobalSuppressions.cs [hunk 2]
```diff
@@ -24,14 +24,12 @@
...
-[assembly: SuppressMessage("Build", "CA1801:Parameter console of method GetMsBuildDirectoryInternal is never used. Remove the parameter or use it in the method body.", Justification = "<Pending>", Scope = "member", Target = "~M:NuGet.CommandLine.MsBuildUtility.GetMsBuildDirectoryInternal(System.String,NuGet.CommandLine.IConsole,System.Collections.Generic.IEnumerable{NuGet.CommandLine.MsBuildToolset},System.Func{NuGet.Common.IEnvironmentVariableReader,System.String})~NuGet.CommandLine.MsBuildToolset")]
+[assembly: SuppressMessage("Build", "CA1801:Parameter console of method GetMsBuildDirectoryInternal is never used. Remove the parameter or use it in the method body.", Justification = "Shipped public API", Scope = "member", Target = "~M:NuGet.CommandLine.MsBuildUtility.GetMsBuildDirectoryInternal(System.String,NuGet.CommandLine.IConsole,System.Collections.Generic.IEnumerable{NuGet.CommandLine.MsBuildToolset},System.Func{NuGet.Common.IEnvironmentVariableReader,System.String})~NuGet.CommandLine.MsBuildToolset")]
```
Comment (kartheekp-ms) [inline src/NuGet.Clients/NuGet.CommandLine/GlobalSuppressions.cs]: parameter `console` is already removed https://github.com/NuGet/NuGet.Client/pull/4177/files#diff-c3438f9ee8b1e6c0ea4470bda47bf8219a8a97dcc3098a172cb51e42451c368fL568. I think we can delete this justification.

File: src/NuGet.Clients/NuGet.CommandLine/GlobalSuppressions.cs [hunk 3]
```diff
@@ -53,7 +51,7 @@
...
-[assembly: SuppressMessage("Build", "CA1801:Parameter registryKeyPath of method GetValueFromRegistryKey is never used. Remove the parameter or use it in the method body.", Justification = "<Pending>", Scope = "member", Target = "~M:NuGet.CommandLine.RegistryKeyUtility.GetValueFromRegistryKey(System.String,System.String,Microsoft.Win32.RegistryKey,NuGet.Common.ILogger)~System.Object")]
+[assembly: SuppressMessage("Build", "CA1801:Parameter registryKeyPath of method GetValueFromRegistryKey is never used. Remove the parameter or use it in the method body.", Justification = "Shipped Public API", Scope = "member", Target = "~M:NuGet.CommandLine.RegistryKeyUtility.GetValueFromRegistryKey(System.String,System.String,Microsoft.Win32.RegistryKey,NuGet.Common.ILogger)~System.Object")]
```
Comment (kartheekp-ms) [inline src/NuGet.Clients/NuGet.CommandLine/GlobalSuppressions.cs]: I think it is safe to remove this justification because code was modified to use `registryKeyPath` parameter https://github.com/NuGet/NuGet.Client/pull/4177/files#diff-c8d727c218517cf8d521a755241e98ba10d1d6ee3033e4d3672508c6f89a9e53R24

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Converters/MessageLevelToBrushConverter.cs [hunk 4]
```diff
@@ -1,73 +0,0 @@
...
-using System.Globalization;
-using System.Windows;
-using System.Windows.Data;
-using System.Windows.Media;
-using NuGet.ProjectManagement;
-
-namespace NuGet.PackageManagement.UI
-{
-    [ValueConversion(typeof(MessageLevel), typeof(Brush))]
-    internal class MessageLevelToBrushConverter : Freezable, IValueConverter
```
Comment (kartheekp-ms) [inline src/NuGet.Clients/NuGet.PackageManagement.UI/Converters/MessageLevelToBrushConverter.cs:11]: I think this is dead code and hence it has been removed from the product.

File: src/NuGet.Clients/NuGet.PackageManagement.UI/NuGet.PackageManagement.UI.csproj [hunk 5]
```diff
@@ -22,6 +22,7 @@
     <NuGetPackageImportStamp>3b445626</NuGetPackageImportStamp>
     <TargetFrameworkVersion>$(NETFXTargetFrameworkVersion)</TargetFrameworkVersion>
     <AlwaysCompileMarkupFilesInSeparateDomain>false</AlwaysCompileMarkupFilesInSeparateDomain>
+    <GenerateAssemblyInfo>false</GenerateAssemblyInfo>
```
Comment (kartheekp-ms) [inline src/NuGet.Clients/NuGet.PackageManagement.UI/NuGet.PackageManagement.UI.csproj]: Is this change intentional?

File: src/NuGet.Clients/NuGet.Tools/NuGet.Tools.csproj [hunk 6]
```diff
@@ -23,6 +23,7 @@
     <VSSDKTargetPlatformRegRootSuffix>Exp</VSSDKTargetPlatformRegRootSuffix>
     <MinimumVisualStudioVersion>16.0</MinimumVisualStudioVersion>
     <RestoreProjectStyle>PackageReference</RestoreProjectStyle>
+    <GenerateAssemblyInfo>false</GenerateAssemblyInfo>
```
Comment (kartheekp-ms) [inline src/NuGet.Clients/NuGet.Tools/NuGet.Tools.csproj]: Is this change intentional?

File: src/NuGet.Clients/NuGet.VisualStudio.Implementation/GlobalSuppressions.cs [hunk 7]
```diff
@@ -11,15 +11,13 @@
...
-[assembly: SuppressMessage("Build", "CA1801:Parameter provider of method AddFromExtension is never used. Remove the parameter or use it in the method body.", Justification = "<Pending>", Scope = "member", Target = "~M:NuGet.VisualStudio.PreinstalledRepositoryProvider.AddFromExtension(NuGet.Protocol.Core.Types.ISourceRepositoryProvider,System.String)")]
+[assembly: SuppressMessage("Build", "CA1801:Parameter provider of method AddFromExtension is never used. Remove the parameter or use it in the method body.", Justification = "Shipped public API", Scope = "member", Target = "~M:NuGet.VisualStudio.PreinstalledRepositoryProvider.AddFromExtension(NuGet.Protocol.Core.Types.ISourceRepositoryProvider,System.String)")]
```
Comment (kartheekp-ms) [inline src/NuGet.Clients/NuGet.VisualStudio.Implementation/GlobalSuppressions.cs]: unused parameter `provider` is already deleted from the code in this PR. https://github.com/NuGet/NuGet.Client/pull/4177/files#diff-a6da29e95db056fb704760cee1337bebd1c9f98f8263f27e04510aacc9702cb3R79. I think we can remove this suppression.

File: src/NuGet.Core/NuGet.Commands/MSBuildProjectFactory.cs [hunk 8]
```diff
@@ -57,7 +57,24 @@ public ILogger Logger
 
         public IMachineWideSettings MachineWideSettings { get; set; }
 
+        /// <summary>
+        /// Creates a MSBuild project factory with data from NuGet <see cref="Commands.PackArgs"/> object
+        /// </summary>
+        /// <param name="packArgs">NuGet arguments for creating a package</param>
+        /// <param name="path">Not used</param>
+        /// <returns>A <see cref="MSBuildProjectFactory"/> object</returns>
+        /// <remarks>This methods is used in <see cref="PackCommandRunner.CreateProjectFactory"/> delegate</remarks>
```
Comment (kartheekp-ms) [inline src/NuGet.Core/NuGet.Commands/MSBuildProjectFactory.cs]: Assuming no references to this method in the product code, can we mark this method obsolete?

**Review by kartheekp-ms — COMMENTED** [review]
Great work Fernando. Left few comments to start with.

File: src/NuGet.Core/NuGet.Commands/MSBuildProjectFactory.cs (hunk 8 above)
Comment (dominoFire) [reply src/NuGet.Core/NuGet.Commands/MSBuildProjectFactory.cs]: No. This is a delegate method implemetation of https://github.com/NuGet/NuGet.Client/blob/563216c35ce91f8c2bdb5ca1d102e639107a40b5/src/NuGet.Core/NuGet.Commands/CommandRunners/PackCommandRunner.cs#L24

File: src/NuGet.Clients/NuGet.CommandLine/Commands/ProjectFactory.cs (hunk 1 above)
Comment (dominoFire) [reply src/NuGet.Clients/NuGet.CommandLine/Commands/ProjectFactory.cs]: My mistake. Path is used. See inherited xmldoc

File: src/NuGet.Clients/NuGet.CommandLine/RegistryKeyUtility.cs [hunk 9]
```diff
@@ -9,11 +9,19 @@ namespace NuGet.CommandLine
...
+        /// <param name="registryKeyPath">Registry key path</param>
+        /// <param name="registryKey">Windows Registry Key object</param>
+        /// <param name="logger">Logger to log errors</param>
+        /// <returns>Registry value or null if there is an error</returns>
         public static object GetValueFromRegistryKey(string name, string registryKeyPath, RegistryKey registryKey, ILogger logger)
         {
             try
             {
-                using (var key = registryKey?.OpenSubKey(name))
+                using (var key = registryKey?.OpenSubKey(registryKeyPath))
```
Comment (nkolev92) [inline src/NuGet.Clients/NuGet.CommandLine/RegistryKeyUtility.cs:24]: This is a functional change right?

Have you checked if there are any bugs in this area?

File: src/NuGet.Core/NuGet.Commands/RestoreCommand/Diagnostics/UnexpectedDependencyMessages.cs [hunk 10]
```diff
@@ -225,9 +224,11 @@ public static bool HasMissingLowerBound(VersionRange range)
         }
 
         /// <summary>
-        /// Log upgrade warnings from the graphs.
+        /// Logs upgrade warnings from the graphs.
         /// </summary>
-        public static IEnumerable<RestoreLogMessage> GetDependenciesAboveUpperBounds(List<IndexedRestoreTargetGraph> graphs, ILogger logger)
+        /// <param name="graphs">Flattened restore graphs with error and warning info</param>
+        /// <returns>A collection of log messages</returns>
+        public static IEnumerable<RestoreLogMessage> GetDependenciesAboveUpperBounds(List<IndexedRestoreTargetGraph> graphs)
```
Comment (nkolev92) [inline src/NuGet.Core/NuGet.Commands/RestoreCommand/Diagnostics/UnexpectedDependencyMessages.cs]: Can you make this internal instead?
It's not anywhere outside this class.

File: src/NuGet.Core/NuGet.Commands/RestoreCommand/RestoreTargetGraph.cs [hunk 11]
```diff
@@ -85,16 +85,31 @@ public class RestoreTargetGraph : IRestoreTargetGraph
             ResolvedDependencies = resolvedDependencies;
         }
 
-        public static RestoreTargetGraph Create(IEnumerable<GraphNode<RemoteResolveResult>> graphs, RemoteWalkContext context, ILogger logger, NuGetFramework framework)
+        /// <summary>
```
Comment (nkolev92) [inline src/NuGet.Core/NuGet.Commands/RestoreCommand/RestoreTargetGraph.cs:88]: This looks to be unused in product code? Can we 🔪 it?

File: src/NuGet.Core/NuGet.Commands/RestoreCommand/Utility/BuildAssetsUtils.cs [hunk 12]
```diff
@@ -399,14 +399,24 @@ public static string GetMSBuildFilePathForPackageReferenceStyleProject(PackageSp
...
+        /// </summary>
+        /// <param name="project">Project/package specification</param>
+        /// <param name="assetsFile">project.assets.json model object</param>
+        /// <param name="targetGraphs"></param>
+        /// <param name="repositories">NuGet package sources</param>
+        /// <param name="request">Restore request model</param>
+        /// <param name="assetsFilePath">File Path from loaded <paramref name="assetsFile"/></param>
+        /// <param name="restoreSuccess">Indicates if packages were restored successfully</param>
+        /// <returns>A collection of .props/.targets file contents generated from restored packages</returns>
         public static List<MSBuildOutputFile> GetMSBuildOutputFiles(PackageSpec project,
```
Comment (nkolev92) [inline src/NuGet.Core/NuGet.Commands/RestoreCommand/Utility/BuildAssetsUtils.cs]: Similar logic as some of the other ones. We can probably make the whole thing internal.

**Review by nkolev92 — COMMENTED** [review]
A lot of work has gone into this, but it's gotten kind of long.

I'm wondering if we can split up the PR in 2. 

For example, things that affect public apis, vs things that don't.

File: src/NuGet.Clients/NuGet.CommandLine/GlobalSuppressions.cs (hunk 2 above)
Comment (dominoFire) [reply src/NuGet.Clients/NuGet.CommandLine/GlobalSuppressions.cs]: Removed

File: src/NuGet.Clients/NuGet.CommandLine/GlobalSuppressions.cs (hunk 3 above)
Comment (dominoFire) [reply src/NuGet.Clients/NuGet.CommandLine/GlobalSuppressions.cs]: Removed.

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Converters/MessageLevelToBrushConverter.cs (hunk 4 above)
Comment (dominoFire) [reply src/NuGet.Clients/NuGet.PackageManagement.UI/Converters/MessageLevelToBrushConverter.cs:11]: Removed.

File: src/NuGet.Clients/NuGet.VisualStudio.Implementation/GlobalSuppressions.cs (hunk 7 above)
Comment (dominoFire) [reply src/NuGet.Clients/NuGet.VisualStudio.Implementation/GlobalSuppressions.cs]: Removed.

File: src/NuGet.Clients/NuGet.PackageManagement.UI/NuGet.PackageManagement.UI.csproj (hunk 5 above)
Comment (dominoFire) [reply src/NuGet.Clients/NuGet.PackageManagement.UI/NuGet.PackageManagement.UI.csproj]: Yes. We have an AssemblyInfo.cs in the project. Sometimes, when you build/rebuild from VS, it ends up with duplicated AssemblyInfo entries 

https://github.com/NuGet/NuGet.Client/blob/563216c35ce91f8c2bdb5ca1d102e639107a40b5/src/NuGet.Clients/NuGet.PackageManagement.UI/NuGet.PackageManagement.UI.csproj#L230

File: src/NuGet.Clients/NuGet.Tools/NuGet.Tools.csproj (hunk 6 above)
Comment (dominoFire) [reply src/NuGet.Clients/NuGet.Tools/NuGet.Tools.csproj]: Yes.

https://github.com/NuGet/NuGet.Client/blob/dev/src/NuGet.Clients/NuGet.Tools/Properties/AssemblyInfo.cs

File: src/NuGet.Clients/NuGet.PackageManagement.UI/NuGet.PackageManagement.UI.csproj (hunk 5 above)
Comment (dominoFire) [reply src/NuGet.Clients/NuGet.PackageManagement.UI/NuGet.PackageManagement.UI.csproj]: Removed, since other builds are unnafected.

File: src/NuGet.Core/NuGet.Commands/MSBuildProjectFactory.cs (hunk 8 above)
Comment (dominoFire) [reply src/NuGet.Core/NuGet.Commands/MSBuildProjectFactory.cs]: Removed, since other builds are unnafected.

File: src/NuGet.Clients/NuGet.Tools/NuGet.Tools.csproj (hunk 6 above)
Comment (dominoFire) [reply src/NuGet.Clients/NuGet.Tools/NuGet.Tools.csproj]: Restored to original form.

File: src/NuGet.Clients/NuGet.CommandLine/RegistryKeyUtility.cs (hunk 9 above)
Comment (dominoFire) [reply src/NuGet.Clients/NuGet.CommandLine/RegistryKeyUtility.cs:24]: This might be related to path too long errors in NuGet.exe ; currently, this only affect in path loo long specific error logging.

Comment (dominoFire) [reply src/NuGet.Clients/NuGet.CommandLine/RegistryKeyUtility.cs:24]: Also, I added an unit test to prove correct functionality.

File: src/NuGet.Core/NuGet.Commands/RestoreCommand/RestoreTargetGraph.cs (hunk 11 above)
Comment (dominoFire) [reply src/NuGet.Core/NuGet.Commands/RestoreCommand/RestoreTargetGraph.cs:88]: It has 3 NuGet test references. I coudn't find any reference inside cross team codebase. Do you mean to dig deeper and remove those references as well?

Comment (nkolev92) [reply src/NuGet.Core/NuGet.Commands/RestoreCommand/RestoreTargetGraph.cs:88]: I think I meant to say it's not used in product code. 

I think things that aren't used in product code would be great candidates to make internal at this poitn. Especially things are just for utility, rather than necessity.

File: src/NuGet.Core/NuGet.Commands/RestoreCommand/Utility/BuildAssetsUtils.cs (hunk 12 above)
Comment (dominoFire) [reply src/NuGet.Core/NuGet.Commands/RestoreCommand/Utility/BuildAssetsUtils.cs]: OK. For now, I made this method internal.

**Comment by dominoFire** [comment]
> A lot of work has gone into this, but it's gotten kind of long.
> 
> I'm wondering if we can split up the PR in 2.
> 
> For example, things that affect public apis, vs things that don't.

Ok, I will split this PR in 4 PRs:

- Code in src\NuGet.Clients - Private APIs
- Code in src\NuGet.Clients - Public APIs
- Code in src\NuGet.Core - Private APIs
- Code in src\NuGet.Core - Public APIs

This PR will be in draft mode and updated when necessary.

**Comment by dominoFire** [comment]
Closing since all patches have been applied. Public APIs will be handled in another way.
